classes/auto_login.php

```diff
@@ -0,0 +1,212 @@
+                    return;
+                }
+                break;
+        }
+
+        // Just as a final check (now we're definitely going to try login), is the plugin actually
+        // configured? It probably is since they set up the autologin option, but...
+        if (!$auth->is_configured()) {
+            return;
+        }
+
+        // Record in session that we attempted login.
+        switch ($auth->config->autologin) {
+            case saml2_settings::OPTION_AUTO_LOGIN_SESSION:
+                $SESSION->auth_saml2_triedautologin = true;
+                break;
+
+            case saml2_settings::OPTION_AUTO_LOGIN_COOKIE:
+                $SESSION->auth_saml2_lastautologincookie = $currentcookie;
+                break;
+        }
+        $SESSION->auth_saml2_autologinattempt = true;
+
+        // Now actually try to log in!
+        self::login($auth);
+    }
+
+    /**
+     * Tries to auto-login by making a passive login request.
+     *
+     * Also called once redirected back after a successful request.
+     *
+     * @param \auth_plugin_saml2 $auth Auth plugin
+     */
+    protected static function login(\auth_plugin_saml2 $auth) {
+        global $CFG, $FULLME, $SESSION, $saml2auth;
+
+        require(__DIR__ . '/../setup.php');
+        $auth = get_auth_plugin('saml2');
+
+        // Set the default IdP to be the first in the list. Used when dual login is disabled.
+        $arr = array_reverse($saml2auth->metadataentities);
+        $metadataentities = array_pop($arr);
+        $idpentity = array_pop($metadataentities);
+        $idp = md5($idpentity->entityid);
+        $SESSION->saml2idp = $idp;
+
+        $simplesaml = new \SimpleSAML\Auth\Simple($auth->spname);
+
+        $params = [
+            'isPassive' => true,
+            'ErrorURL' => $CFG->wwwroot . '/auth/saml2/autologin.php?success=0&url=' .
+                    urlencode((new \moodle_url($FULLME))->out_as_local_url(false)),
+            'ReturnTo' => $CFG->wwwroot . '/auth/saml2/autologin.php?success=1&url=' .
+                urlencode((new \moodle_url($FULLME))->out_as_local_url(false))
+        ];
+
+        $simplesaml->requireAuth($params);
+
+        // We only get back here if they already logged in.
+        $attributes = $simplesaml->getAttributes();
+        $auth->saml_login_complete($attributes);
+    }
+
+    /**
+     * Finishes the login (called from autologin.php).
+     *
+     * This function redirects if successful.
+     *
+     * @param bool $success True if successful login
+     * @param \moodle_url $url URL for redirecting to
+     */
+    public static function finish($success, \moodle_url $url) {
+        global $SESSION;
+        if (empty($SESSION->auth_saml2_autologinattempt)) {
+            return;
+        }
+        unset($SESSION->auth_saml2_autologinattempt);
+
+        // If they are already logged in, we don't need autologin.
```

Comment:
seems like lines 181-201 are more or less the same in self::consider method. Should we abstract those lines to reduce code duplication?

Reply:
I've broken out two of the checks into a function, we could move the is_configured check in there as well if you think the performance thing isn't worth retaining.